// @streamrpg/shared/src/xp.ts
export const MAX_LEVEL = 30;
export const XP_PER_PING = 10;
export const GOLD_PER_PING = 0.3;
export const PING_COOLDOWN_MS = 60_000;
export const DROP_CHANCE = 0.008;
export const OVERLAY_ACTIVE_SECONDS = 300;

export function xpForLevel(level: number): number {
  return Math.floor(100 * Math.pow(level, 1.5));
}

function buildXpTable(): number[] {
  const table: number[] = [0];
  for (let lvl = 1; lvl < MAX_LEVEL; lvl++) {
    table.push(table[lvl - 1] + xpForLevel(lvl));
  }
  return table;
}

const XP_TABLE = buildXpTable();

export function getLevel(totalXp: number): number {
  let level = 1;
  for (let i = 0; i < MAX_LEVEL - 1; i++) {
    if (totalXp >= XP_TABLE[i + 1]) {
      level = i + 2;
    } else {
      break;
    }
  }
  return level;
}

export interface XpProgress {
  level: number;
  xp: number;
  xp_to_next: number;
  percent: number;
}

export function getProgress(totalXp: number): XpProgress {
  const level = getLevel(totalXp);

  if (level >= MAX_LEVEL) {
    return { level: MAX_LEVEL, xp: 0, xp_to_next: 0, percent: 100 };
  }

  const baseXp = XP_TABLE[level - 1];
  const xpInLevel = totalXp - baseXp;
  const xpNeeded = xpForLevel(level);

  return {
    level,
    xp: xpInLevel,
    xp_to_next: xpNeeded - xpInLevel,
    percent: Math.min(100, Math.floor((xpInLevel / xpNeeded) * 100)),
  };
}

export const RARITY_WEIGHTS: Record<string, number> = {
  common: 6000,
  uncommon: 2500,
  rare: 1300,
  epic: 180,
  legendary: 20,
};

export function pickRarity(rng = Math.random()): string {
  const total = Object.values(RARITY_WEIGHTS).reduce((a, b) => a + b, 0);
  let roll = rng * total;
  for (const [rarity, weight] of Object.entries(RARITY_WEIGHTS)) {
    roll -= weight;
    if (roll <= 0) return rarity;
  }
  return "common";
}
